src/contracts/CrowdfundStateful.ts

```diff
@@ -0,0 +1,171 @@
+        this.donorRefunded = donorRefunded
+        this.deadline = deadline
+        this.target = target
+    }
+
+    // method to donate
+    @method()
+    public donate(contributor: PubKey, amount: bigint) {
+        // donation not equal to zero
+        assert(amount > 0n, 'Donation should be greater than 0')
+
+        // can only donate once
+        assert(
+            !this.donor.has(contributor),
+            'donation from this pub key already present'
+        )
+
+        this.donor.set(contributor, amount)
+
+        //setting refunded to false
+        this.donorRefunded.set(contributor, false)
+
+        assert(
+            this.ctx.sequence < CrowdfundStateful.UINT_MAX,
+            'require nLocktime enabled'
+        )
+
+        // Check if using block height.
+        if (this.deadline < CrowdfundStateful.LOCKTIME_BLOCK_HEIGHT_MARKER) {
+            // Enforce nLocktime field to also use block height.
+            assert(
+                this.ctx.locktime <
+                    CrowdfundStateful.LOCKTIME_BLOCK_HEIGHT_MARKER
+            )
+        }
+        assert(
+            this.ctx.locktime >= this.deadline,
+            ' You cannot donate afterfundraising expired'
+        )
+
+        //updating the contract state
+        let output: ByteString = this.buildStateOutput(
+            this.ctx.utxo.value + amount
+        )
+
+        // handling change output
+        if (this.changeAmount > 0n) {
+            output += this.buildChangeOutput()
+        }
+
+        assert(hash256(output) == this.ctx.hashOutputs, 'hashOutputs mismatch')
+    }
+
+    // Method to collect pledged fund.
+    @method()
+    public collect(sig: Sig) {
+        // Ensure the collected amount actually reaches the target.
+        assert(this.ctx.utxo.value >= this.target)
+
+        // Funds go to the beneficiary.
+        let outputs = Utils.buildPublicKeyHashOutput(
+            hash160(this.beneficiary),
+            this.ctx.utxo.value
+        )
+
+        // Handling change output.
+        if (this.changeAmount > 0n) {
+            outputs += this.buildChangeOutput()
+        }
+
+        assert(hash256(outputs) == this.ctx.hashOutputs, 'hashOutputs mismatch')
+        // Validate signature of beneficiary
+        assert(
+            this.checkSig(sig, this.beneficiary),
+            'beneficiary signature check failed'
+        )
+    }
+
+    // donors can be refunded after the deadline.
+    @method()
+    public refund(contributor: PubKey, amount: bigint, sig: Sig) {
+        assert(this.donor.canGet(contributor, amount))
+
+        // make sure it has not refunded before
+        assert(
+            this.donorRefunded.canGet(contributor, false),
+            'already refunded'
+        )
+
+        // update state
+        const output: ByteString = this.buildStateOutput(
+            this.ctx.utxo.value - amount
+        )
+
+        // setting it as refunded
+        this.donorRefunded.set(contributor, true)
```

Comment:
move this line before `this.buildStateOutput`

it's essential to finish updating all the state props before calling `this.buildStateOutput`

Reply:
Ok will do that